python/GafferSceneTest/InstancerTest.py

```diff
@@ -649,6 +649,355 @@ def testIndices( self ) :
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/bar/{i}/baz/cube".format( i=i ) ), script["cube"]["out"].object( "/cube" ) )
+
+		for i in [ "1", "2" ] :
+
+			self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/foo/{i}".format( i=i ) ), IECore.InternedStringVectorData( [ "bar" ] ) )
+			self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/foo/{i}/bar".format( i=i ) ), IECore.InternedStringVectorData( [ "sphere" ] ) )
+
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/foo/{i}".format( i=i ) ), IECore.NullObject.defaultNullObject() )
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/foo/{i}/bar".format( i=i ) ), IECore.NullObject.defaultNullObject() )
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/foo/{i}/bar/sphere".format( i=i ) ), script["sphere"]["out"].object( "/sphere" ) )
+
+		self.assertSceneValid( script["instancer"]["out"] )
+
+	def assertRootsToLeaves( self, script ) :
+
+		self.assertEqual( script["instancer"]["out"].childNames( "/object/instances" ), IECore.InternedStringVectorData( [ "sphere", "cube" ] ) )
+		self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/sphere" ), IECore.InternedStringVectorData( [ "0", "3" ] ) )
+		self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/cube" ), IECore.InternedStringVectorData( [ "1", "2" ] ) )
+
+		self.assertEqual( script["instancer"]["out"].object( "/object/instances" ), IECore.NullObject.defaultNullObject() )
+		self.assertEqual( script["instancer"]["out"].object( "/object/instances/sphere" ), IECore.NullObject.defaultNullObject() )
+		self.assertEqual( script["instancer"]["out"].object( "/object/instances/cube" ), IECore.NullObject.defaultNullObject() )
+
+		for i in [ "0", "3" ] :
+
+			self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/sphere/{i}".format( i=i ) ), IECore.InternedStringVectorData( [] ) )
+
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/sphere/{i}".format( i=i ) ), script["sphere"]["out"].object( "/sphere" ) )
+
+		for i in [ "1", "2" ] :
+
+			self.assertEqual( script["instancer"]["out"].childNames( "/object/instances/cube/{i}".format( i=i ) ), IECore.InternedStringVectorData( [] ) )
+
+			self.assertEqual( script["instancer"]["out"].object( "/object/instances/cube/{i}".format( i=i ) ), script["cube"]["out"].object( "/cube" ) )
+
+		self.assertSceneValid( script["instancer"]["out"] )
+
+	def testIndexedRootsList( self ) :
+
+		script = self.buildPrototypeRootsScript()
+		script["instancer"]["prototypeMode"].setValue( GafferScene.Instancer.PrototypeMode.IndexedRootsList )
+
+		script["instancer"]["prototypeRootPaths"].setValue( IECore.StringVectorData( [] ) )
+		self.assertRootsMatchPrototypeSceneChildren( script )
+
+		script["instancer"]["prototypeRootPaths"].setValue( IECore.StringVectorData( [ "", ] ) )
```

Comment:
Sure, sounds fine. One can always set "/" explicitly to get the other behaviour. To clarify what you're afeter, if the rootpaths are set to `[ "/foo", "", "/bar" ]` are you expecting all points with `index==1` to be skipped? Or were you expecting an implicit index-shift (eg they'd all get "/bar")? I think I favor skipping.

Reply:
Skipping sounds good to me.

Reply:
I started implementing this, but I think its going to complicate `Engine::prototypeIndex`. Some points will now have "invalid" indices, so we need to validate them rather than just fetch from `m_indices`. Do you see a simpler way? If not, is the added complexity worth it or should we just handle the trailing "" case (eg `[ "" ]` and `[ "/foo", "/bar", "" ]`)?

Reply:
That makes sense, but I think it's worth the extra work to make things bombproof. Seems like it shouldn't be _too_ hard? I would imagine that in `initPrototypes()` we need to initialise a `vector<int> m_prototypeIndexToValidatedPrototypeIndex` (for want of a better name), where we store -1 for invalid indices and adjust the other indices to account for gaps. And then just index into that in `prototypeIndex()`, and skip the invalid indices in the compute for `prototypeChildNamesPlug()`. Did I understand that right?

Reply:
Yeah, it'd be something like that. I'll give it a go then.

Dan and I talked yesterday and thought there might be some way to avoid it if ChildNamesMap had a way to increment the index (ie. track validity itself). But I'm not sure I want to get into all that right now...

Reply:
Yeah, it's nice that ChildNamesMap has been helpful, but it is designed for a slightly different use case, so I think we're better off keeping the special case part in the Instancer where it's needed.

Reply:
Addressed in ac8696aad81f52bb6998892fbbaad8e35cc2a5cd

I also added afee26c44ea24be2cc1f11985f4e27ec2ac4f006 to allow the "/" option we described in this discussion. Its unfortunate that it needs a special case in the code, but I think it is useful. I'm not at all attached to using "root" as the name though... Do you have a better suggestion?